File: streamkv-java/src/main/java/streamkv/api/java/KVStore.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package streamkv.api.java;

import java.util.ArrayList;
import java.util.List;

import org.apache.flink.api.common.ExecutionConfig;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.IterativeDataStream;

import streamkv.api.java.kvstorebuilder.JavaKVStoreBuilder;
import streamkv.api.java.kvstorebuilder.AbstractKVStoreBuilder;
import streamkv.api.java.util.KVUtils;

/**
 * Basic streaming key-value store abstraction. The user can use Flink
 * {@link DataStream}s to apply operations on the store, such as {@link #put} or
 * {@link #get}. Operations that generate output return a query id that can be
 * used to retrieve the result streams from the {@link KVStoreOutput} instance
 * after calling {@link #getOutputs()}.
 * 
 * <p>
 * Calling {@link KVStore#getOutputs()} will finalize the operations applied to
 * the store, and no further operations can be applied after that in order to
 * avoid creating circular dependencies in the resulting program. If such logic
 * is necessary, it needs to be handled manually using
 * {@link IterativeDataStream}.
 * </p>
 * 
 * @param <K>
 *            Type of the keys.
 * @param <V>
 *            Type of the values.
 */
public class KVStore<K, V> {

	private final AbstractKVStoreBuilder<K, V> storeBuilder;
	private List<Query<?>> queries = new ArrayList<>();

	private KVStore(OperationOrdering ordering) {
		this.storeBuilder = new JavaKVStoreBuilder<>(ordering);
	}

	/**
	 * Put the elements from the given tuple stream into the key-value store.
	 * 
	 * @param stream
	 *            Stream of {@link Tuple2}s representing the (K,V) pairs.
	 */
	public void put(DataStream<Tuple2<K, V>> stream) {
		int qid = storeBuilder.nextID();
		storeBuilder.put(stream.map(new KVUtils.ToPut<K, V>(qid)), qid);
	}

	/**
	 * Update the value of the provided key by reducing it with the current
	 * value using the reduce function provided. If there is no entry update
	 * this works like a put operation and creates a new KV entry for the given
	 * key.
	 * 
	 * @param stream
	 *            Stream of {@link Tuple2}s representing the (K,V) pairs.
	 * @param reducer
	 *            The reducer used to update the elements.
	 */
	public void update(DataStream<Tuple2<K, V>> stream, ReduceFunction<V> reducer) {
		int qid = storeBuilder.nextID();
		storeBuilder.update(stream.map(new KVUtils.ToUpdate<K, V>(qid)), reducer, qid);
	}

	/**
	 * Get elements from the store by specifying a stream of keys to retrieve.
	 * 
	 * @param stream
	 *            The stream of keys to get.
	 * @return The resulting {@link Query}.
	 */
	public Query<Tuple2<K, V>> get(DataStream<K> stream) {
		int qid = storeBuilder.nextID();
		storeBuilder.get(stream.map(new KVUtils.ToGet<K, V>(qid)), qid);
		Query<Tuple2<K, V>> q = new Query<Tuple2<K, V>>(qid, storeBuilder);
		queries.add(q);
		return q;
	}

	/**
	 * Remove elements from the store by specifying a stream of keys to remove.
	 * 
	 * @param stream
	 *            The stream of keys to remove.
	 * @return The resulting {@link Query}.
	 */
	public Query<Tuple2<K, V>> remove(DataStream<K> stream) {
		int qid = storeBuilder.nextID();
		storeBuilder.remove(stream.map(new KVUtils.ToRemove<K, V>(qid)), qid);
		Query<Tuple2<K, V>> q = new Query<Tuple2<K, V>>(qid, storeBuilder);
		queries.add(q);
		return q;
	}

	/**
	 * Get elements from the store by specifying a stream of records and a
	 * {@link KeySelector} for extracting the key from each record.
	 * 
	 * @param stream
	 *            The stream of records for which the key will be extracted.
	 * @param keySelector
	 *            The {@link KeySelector} used to extract the key for each
	 *            element.
	 * @return The resulting {@link Query}.
	 */
	@SuppressWarnings({ "unchecked", "rawtypes" })
	public <X> Query<Tuple2<X, V>> getWithKeySelector(DataStream<X> stream, KeySelector<X, K> keySelector) {
		int qid = storeBuilder.nextID();
		storeBuilder.selectorGet(((DataStream) stream).map(new KVUtils.ToSGet<K, V>(qid)), keySelector, qid);
		Query<Tuple2<X, V>> q = new Query<Tuple2<X, V>>(qid, storeBuilder);
		queries.add(q);
		return q;
	}

	/**
	 * Get multiple elements from the store at the same time by specifying a
	 * stream of key arrays to retrieve.
	 * 
	 * @param stream
	 *            The stream of key arrays to get.
	 * @return The resulting {@link Query}.
	 */
	public Query<Tuple2<K, V>[]> multiGet(DataStream<K[]> stream) {
		int qid = storeBuilder.nextID();
		storeBuilder.multiGet(stream.flatMap(new KVUtils.ToMGet<K, V>(qid)), qid);
		Query<Tuple2<K, V>[]> q = new Query<Tuple2<K, V>[]>(qid, storeBuilder);
		queries.add(q);
		return q;
	}

	/**
	 * Get multiple elements from the store at the same time by specifying a
	 * stream of object arrays and a {@link KeySelector} for extracting the keys
	 * from the objects.
	 * 
	 * @param stream
	 *            The stream of object arrays.
	 * @param keySelector
	 *            The {@link KeySelector} used to extract the key for each
	 *            element.
	 * @return The resulting {@link Query}.
	 */
	@SuppressWarnings({ "unchecked", "rawtypes" })
	public <X> Query<Tuple2<X, V>[]> multiGetWithKeySelector(DataStream<X[]> stream,
			KeySelector<X, K> keySelector) {
		int qid = storeBuilder.nextID();
		storeBuilder.selectorMultiGet(((DataStream) stream).flatMap(new KVUtils.ToSMGet<K, V>(qid)),
				keySelector, qid);
		Query<Tuple2<X, V>[]> q = new Query<Tuple2<X, V>[]>(qid, storeBuilder);
		queries.add(q);
		return q;
	}

	/**
	 * Returns a list of all {@link Query}s applied on this store.
	 * 
	 */
	public List<Query<?>> getQueries() {
		return queries;
	}

	/**
	 * Creates a new {@link KVStore} with the given {@link OperationOrdering}
	 * semantics. <br>
	 * <br>
	 * Currently there are 2 supported ordering semantics:
	 * <ul>
	 * <li><b>ARRIVALTIME</b> : All operations are executed in arrival order
	 * (governed by the standard Flink partial ordering guarantees). While this
	 * implementation provides maximal performance it does not provide any
	 * deterministic processing guarantee.</li>
	 * <li><b>TIMESTAMP</b> : All operations are executed in time order. Time can be
	 * ingress time by default or custom event timestamps and watermarks must be
	 * provided by the source implementations. There is no ordering guarantee
	 * among elements with the same timestamps.
	 * 
	 * <p>
	 * This implementation provides deterministic processing guarantees given
	 * that each record has a unique timestamp.
	 * </p>
	 * 
	 * <p>
	 * Record timestamps need to be enabled by calling
	 * {@link ExecutionConfig#enableTimestamps()}.
	 * </p>
	 * </li>
	 * </ul>
	 * 
	 * @param ordering
	 *            {@link OperationOrdering} semantics for the {@link KVStore}.
	 * @return A new {@link KVStore} instance.
	 */
	public static <K, V> KVStore<K, V> withOrdering(OperationOrdering ordering) {
		return new KVStore<K, V>(ordering);
	}
}

File: streamkv-scala/src/main/scala/streamkv/api/scala/KVStore.scala
/*
 * Copyright 2015 Gyula Fóra
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package streamkv.api.scala

import org.apache.flink.streaming.api.scala.DataStream
import streamkv.api.java.{ KVStore => JavaStore }
import streamkv.api.java.OperationOrdering
import streamkv.api.java.types.KVOperation
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator
import org.apache.flink.api.common.functions.ReduceFunction
import scala.util.Random
import org.apache.flink.api.java.functions.KeySelector
import scala.reflect.ClassTag
import org.apache.flink.api.common.typeinfo.TypeInformation
import streamkv.api.scala.kvstorebuilder.ScalaStoreBuilder
import streamkv.api.scala.scala.kvstorebuilder.ScalaStoreBuilder
import scala.collection.mutable.MutableList

import _root_.scala.reflect.ClassTag
import _root_.scala.util.Random

trait KVStore[K, V] {
  /**
   * Put the elements from the given tuple into the key-value store.
   */
  def put(stream: DataStream[(K, V)])
  /**
   * Update the current value for the given key-value pair, by reducing the old and new value with the reduce function.
   */
  def update(stream: DataStream[(K, V)])(reducer: (V, V) => V)
  /**
   * Get elements from the KVStore by key.
   */
  def get(stream: DataStream[K]): Query[(K, V)]
  /**
   * Remove elements from the KVStore by key.
   */
  def remove(stream: DataStream[K]): Query[(K, V)]
  /**
   * Get multiple elements at the same type using an array of keys.
   */
  def multiGet(stream: DataStream[Array[K]]): Query[Array[(K, V)]]
  /**
   * Get elements from the KVStore by passing an object and a function that extracts the key from the object.
   */
  def getWithKeySelector[R](stream: DataStream[R])(key: R => K): Query[(R, V)]
  /**
   * Get multiple elements at the same type using an array of objects and a custom key extractor.
   */
  def multiGetWithKeySelector[R](stream: DataStream[Array[R]])(key: R => K): Query[Array[(R, V)]]
  /**
   * Return all the queries that result in an output stream.
   */
  def getQueries: List[Query[_]]
}

class ScalaKVStore[K: TypeInformation: ClassTag, V: TypeInformation: ClassTag](ordering: OperationOrdering) extends KVStore[K, V] {
  val storeBuilder = new ScalaStoreBuilder[K, V](ordering)
  val queries = new MutableList[Query[_]]

  def put(stream: DataStream[(K, V)]) = {
    val qid = storeBuilder.nextID
    val opstream = stream.map(in => KVOperation.put(qid, in._1, in._2))
      .getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    storeBuilder.put(opstream, qid)
  }
  def update(stream: DataStream[(K, V)])(reduceFun: (V, V) => V) = {
    val qid = storeBuilder.nextID
    val opstream = stream.map(in => KVOperation.update(qid, in._1, in._2))
      .getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    val reducer = new ReduceFunction[V] {
      def reduce(v1: V, v2: V) = { reduceFun(v1, v2) }
    }
    storeBuilder.update(opstream, reducer, qid)
  }
  def get(stream: DataStream[K]) = {
    val qid = storeBuilder.nextID
    val opstream = stream.map(in => KVOperation.get(qid, in))
      .getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    storeBuilder.get(opstream, qid)
    val q = new Query[(K, V)](qid, storeBuilder)
    queries += q
    q
  }
  def remove(stream: DataStream[K]) = {
    val qid = storeBuilder.nextID
    val opstream = stream.map(in => KVOperation.remove(qid, in))
      .getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    storeBuilder.remove(opstream, qid)
    val q = new Query[(K, V)](qid, storeBuilder)
    queries += q
    q
  }
  def multiGet(stream: DataStream[Array[K]]) = {
    val qid = storeBuilder.nextID
    val opstream = stream.flatMap(in => {
      val opID = Random.nextLong
      val numKeys: Short = in.length.toShort
      for (i <- 0 to numKeys - 1) yield KVOperation.multiGet(qid, in(i), numKeys, opID)
    }).getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    storeBuilder.multiGet(opstream, qid)
    val q = new Query[Array[(K, V)]](qid, storeBuilder)
    queries += q
    q
  }
  def getWithKeySelector[R](stream: DataStream[R])(key: R => K) = {
    val qid = storeBuilder.nextID
    val opstream = stream.map(in => KVOperation.selectorGet(qid, in))
      .getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    val selector = new KeySelector[R, K] {
      def getKey(o: R) = key(o)
    }
    storeBuilder.selectorGet(opstream, selector, qid)
    val q = new Query[(R, V)](qid, storeBuilder)
    queries += q
    q
  }

  def multiGetWithKeySelector[R](stream: DataStream[Array[R]])(key: R => K) = {
    val qid = storeBuilder.nextID
    val opstream = stream.flatMap(in => {
      val opID = Random.nextLong
      val numKeys: Short = in.length.toShort
      for (i <- 0 to numKeys - 1) yield KVOperation.selectorMultiGet(qid, in(i), numKeys, opID)
    }).getJavaStream.asInstanceOf[SingleOutputStreamOperator[KVOperation[K, V], _]]
    val selector = new KeySelector[R, K] {
      def getKey(o: R) = key(o)
    }
    storeBuilder.selectorMultiGet(opstream, selector, qid)
    val q = new Query[Array[(R, V)]](qid, storeBuilder)
    queries += q
    q
  }

  def getQueries: List[Query[_]] = queries.toList
}

object KVStore {
  def apply[K: TypeInformation: ClassTag, V: TypeInformation: ClassTag](ordering: OperationOrdering): KVStore[K, V] = new ScalaKVStore[K, V](ordering)
}